thriftbp/server_middlewares.go

```diff
@@ -477,3 +484,32 @@ func PrometheusServerMiddleware(method string, next thrift.TProcessorFunction) t
 	}
 	return thrift.WrappedTProcessorFunction{Wrapped: process}
 }
+
+// EnvoyGracefulDrainHeader returns middleware which sets a drain signal header on server responses.
+func EnvoyGracefulDrainHeader() thrift.ProcessorMiddleware {
+
+	const envoyDrainHeader = ":drain"
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
+	var shouldDrain int32
+	go func() {
+		<-sig
+		atomic.StoreInt32(&shouldDrain, 1)
+	}()
```

Comment:
I would slightly prefer to make a change to `runtimebp` package to provide this global singleton instead of reimplement it here.

my other concern is that, can you elaborate on what real difference this makes on the envoy level? is it to make sure that envoy does not send more requests to this server pod? the current implementation (the one not using global singleton) can totally be implemented in service code instead, can you implement this in your service first, test it in staging/run it in production, fix bugs/verify that it does what we expect it does, then upstream it to baseplate?

Reply:
You can wire function calls into both the pre and post stop steps in baseplate.go (I think this would need to be in pre-stop). https://pkg.go.dev/github.com/reddit/baseplate.go#ServeArgs

Rather than using a global state, you could make a struct that stores the state and has methods for the shutdown handler and the middleware.

```go
const envoyDrainHeader = ":drain"

type EnvoyGracefulDrainHeader struct {
	shouldDrain *int32
}

func (e *EnvoyGracefulDrainHeader) Close() error {
	atomic.StoreInt32(e.shouldDrain, 1)	
}

func (e *EnvoyGracefulDrainHeader) ThriftProcessorMiddleware(name string, next thrift.TProcessorFunction) thrift.TProcessorFunction {
	return thrift.WrappedTProcessorFunction{
		Wrapped: func(ctx context.Context, seqID int32, in, out thrift.TProtocol) (ok bool, err thrift.TException) {
			if atomic.LoadInt32(e.shouldDrain) > 0 {
				if t, ok := out.(*thrift.THeaderProtocol); ok {
					t.SetWriteHeader(envoyDrainHeader, "true")
				}
			}

			return next.Process(ctx, seqID, in, out)
		},
	}
}
```

Reply:
> I would slightly prefer to make a change to runtimebp package to provide this global singleton instead of reimplement it here.

This also sounds like a reasonable approach, but maybe it should be on the `Baseplate` object rather than a global? Reason being is the function in runtimebp is really just a helper, while I think this notion of the server running belongs in baseplate.Serve. Maybe add a `ServerIsRunning() bool` function to the Baseplate?